feat: support static reference in LibraryContentBlock

connorhaugh opened the pull request:
Requires:

* https://github.com/openedx/edx-platform/pull/33263

## Description

This PR adds the ability for course authors to reference specific pieces of content from a content library in their course using the library xblock. It does so for v1 and v2 xblocks.

It does so by implementing the changes described in the [ADR which describes the new desired schema for the block](https://github.com/openedx/edx-platform/pull/33231)

## Testing instructions

To test this PR:

1. Create a v1 and a v2 content library, "v1 lib" and "v2 lib." Make sure each has some problems in it.
2. In a unit in a course, create 2 library blocks. associate v1 lib with the first block, and v2 lib with the second.
  
5. Test Randomized content behavior: with shuffle = true and max_count = 1, save and publish the blocks.
expected: you should see 1 block at random per user per block. Try again with max_count = 2, and then max_count =  -1. You should then see 2 and all blocks respectively. 

7. Test Specific content reference behavior. With shuffle = false and manual = true, select one piece of content from the v1 and v2 libs. Republish. This content should then be displayed to the user.

kdmccormick committed:
feat!: remove LibrarySourcedBlock

Rather than implementing V2-library and static-library-reference support
in a new block, we will be enhancing the existing `LibraryContentBlock`
in-place.

Relevant ADR PR: https://github.com/openedx/edx-platform/pull/33231

kdmccormick committed:
feat: support V2 libraries in LibraryContentBlock (randomized only)

Co-Authored-By: Connor Haugh <chaugh@2u.com>
Co-Authored-By: Eugene Dyudyunov <evgen.dyudyunov@raccoongang.com>

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

Commit:
fix: user tasks out of lms, move tasks.py

Commit:
Merge commit 'refs/pull/33263/head' of github.com:openedx/edx-platform into kdmccormick/library-content-v2

Commit:
fix: lint

Commit:
fix: lint

Commit:
fix: lint fix

Commit:
fix: move codeowner attribute to layer

Commit:
feat: implement chema

kdmccormick committed:
fix: remove codeowner attribute from update_children_task

...per comment in cms/djangoapps/contentstore/tasks.py.

This fixes an exception that was raised in the background whenever
the update_children_task was started:

     Traceback (most recent call last):
       File "/openedx/venv/lib/python3.8/site-packages/celery/utils/dispatch/signal.py", line 276, in send
         response = receiver(signal=self, sender=sender, **named)
       File "/openedx/venv/lib/python3.8/site-packages/user_tasks/signals.py", line 215, in start_user_task
         sender.status.start()
       File "/openedx/venv/lib/python3.8/site-packages/user_tasks/tasks.py", line 84, in status
         name = self.generate_name(arguments_dict)
       File "/openedx/edx-platform/openedx/core/djangoapps/content_libraries/tasks.py", line 274, in generate_name
         key = arguments_dict['dest_block_key']
     KeyError: 'dest_block_key'

kdmccormick committed:
build: put back set_code_owner_attribute but as a regular function call

kdmccormick committed:
feat: add cms-only assertions to content_libraries tasks

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
style: fix indentation

kdmccormick committed:
fix: add missing imports for _assert_cms

kdmccormick committed:
docs: remove outdated is_v2_lib from function docstring

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
test: run ./xmodule/ tests with CMS settings

Just like ./common/ and ./openedx/, the unit tests in ./xmodule/
validate behavior in both LMS and CMS. In order to fully test
./xmodule/, we need to run its tests in both contexts.

Also in this commit:

* refactor: Rename the shards to be clear whether they're
            running under LMS or CMS.
* docs: correct comments regarding conditions under which
        codejail's test_cant_do_something_forbidden is skipped.

* test: update a unit test which was using the now-deleted
        library_sourced block to use library_content block instead.

kdmccormick committed:
test: run ./xmodule/ LibraryContentTests only in CMS

kdmccormick committed:
test: run ./xmodule/ LibraryContentTests only in CMS

kdmccormick committed:
Revert "test: run ./xmodule/ LibraryContentTests only in CMS"

This reverts commit f1370d0b33674925ed0240567f3fb1ebfcd2a814.

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
Merge branch 'kdmccormick/test-xmodule-with-cms' into kdmccormick/library-content-v2

kdmccormick committed:
style: remove duplicate imports fomr test_lib_tools

kdmccormick committed:
test: fix patching in test_list_available_libraries

kdmccormick committed:
test: attempt at stricter enforcement of library tools in cms only

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
temp: fix: put library_tools back in LMS, at least for now

kdmccormick committed:
test: fix skip_unless_[lms|cms]

kdmccormick committed:
test: fix library_content completion test w/ settings hack

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
test: update test patch for new libraries tasks location

kdmccormick committed:
test: use string mock to satisfy importlinter

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

Commit:
fix: lint + add unit tests

connorhaugh committed:
Merge branch 'kdmccormick/library-content-v2' into feat--Library-Content-Block-Reference

Commit:
fix: code quality cleanup

Commit:
Merge branch 'feat--Library-Content-Block-Reference' of github.com:edx/edx-platform into feat--Library-Content-Block-Reference

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

Commit:
fix: module resolution order for import

Commit:
fix: reset default max_count to 1 from -1

Commit:
fix: correctly pass context vars block selection

connorhaugh commented on xmodule/library_content_block.py:
```diff
@@ -127,17 +142,6 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
         display_name=_("Library Version"),
         scope=Scope.settings,
     )
-    mode = String(
```
Deprecating mode is something I wanted to go back to and clean up. Initially, I was worried about when old olx without the "shuffle" field would be read from storage. I was fearful that it wouldn't be able to parse correctly. After some investigation, however, keeping shuffle's default to be "True" allows for the legacy behavior to be maintained without any effort. I looked into augmenting the "parse_xml" function to see if that could overwrite the "shuffle" value, but that gets messy because if a block has mode set and shuffle = false, what do we rely on as our source of truth. This approach of removal keeps it simple.

kdmccormick replied:
Great!

kdmccormick committed:
fix: getting status of library_content's update_children task

kdmccormick committed:
fix: getting status of library_content's update_children task

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

Commit:
fix: final organization of fields

kdmccormick committed:
fix: squash: _are_children_updating utiltity func

kdmccormick committed:
Revert "fix: squash: _are_children_updating utiltity func"

This reverts commit 535fb52ffd554f38adbdf02d49f124714883dcd8.

kdmccormick committed:
Revert "fix: getting status of library_content's update_children task"

This reverts commit 0cf735fcab69b0c011e1991bc7aa5ea113dc1399.

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

Commit:
fix: clear candidates when changing libraries

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
Merge branch 'master' into kdmccormick/library-content-v2

kdmccormick committed:
refactor: delete unused library_tools code; rename update_children

kdmccormick committed:
test: update unit tests for previous refactor commit

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into kdmccormick/library-content-v2

kdmccormick committed:
fix: duplicate library_content children asynchronously (#33652)

connorhaugh committed:
Merge branch 'kdmccormick/library-content-v2' into feat--Library-Content-Block-Reference

Commit:
fix: remove print statments + fix tests

kdmccormick commented on cms/djangoapps/contentstore/xblock_storage_handlers/view_handlers.py:
```diff
@@ -387,12 +398,11 @@ def _save_xblock(  # lint-amnesty, pylint: disable=too-many-statements
     fields=None,
     summary_configuration_enabled=None,
-):
+):  # lint-amnesty, pylint: disable=too-many-statements
```
if you want to pylint-disable something, do not include lint-amnesty. lint-amnesty is only for violations that were grandfathered in when we started enforcing 0 violations.
```suggestion
):  # pylint: disable=too-many-statements
```

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -97,23 +104,31 @@ class LibraryContentBlock(
+    }
+    studio_view_css = {
+        'scss': [],
+    }
```
```suggestion
    mako_template = 'widgets/metadata-edit.html'
```
This is all old stuff from before the xmodule asset refactoring.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -97,23 +104,31 @@ class LibraryContentBlock(
         display_name=_("Display Name"),
         help=_("The display name for this component."),
-        default="Randomized Content Block",
+        default="Library Reference Block",
```
```suggestion
        default="Libraries Component",
```
I believe @jmakowski1123 and Brad (GH username?) talked and decided that the new name of this component should just be "Libraries". Given that, do you guys have thoughts on what the default title of the component should be? (This is the title that shows up when you first create a Libraries component, but haven't given it a title yet. Currently it's "Randomized Content Block", but we could call it anything.)

kdmccormick replied:
Asked in slack https://openedx.slack.com/archives/C05NT2YN820/p1701714420089529

kdmccormick replied:
@connorhaugh We landed on "Library Content" for the default title, can you make that change here?

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -151,12 +155,29 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
+        display_name=_("Manually Selected Blocks"),
+        default=[],
+        scope=Scope.settings,
+    )
```
@connorhaugh can you remind me why we want this instead of just using `children`? Is it so that if you change the mode to randomized and then change it back to static, we preserve the author's manually chosen blocks?

once we agree, can you update [the adr](https://github.com/openedx/edx-platform/blob/master/xmodule/docs/decisions/0003-library-content-block-schema.rst) with that reasoning?

connorhaugh replied:
So `children` currently refers to the set of all blocks which have been copied into the course.

The blocks selected for specific reference are a strict subset of `children.`

We use `children` presently, therefore, to be the list authors use to select their candidates from.

If we set children=candidates, then authors would not be able to manipulate their selection.

If we tried to pull from the library every time we wanted to create a list for selection, we would lose overrides, etc.

More than happy to update the adr.

kdmccormick replied:
@connorhaugh Got it. That makes sense. I like name `candidates` too.

Two qs, which will affect how the LCB editor behaves when you switch between Random and Manual modes:

* When `manual == False`, will you ignore `candidates`, or will you adjust it to be equal to the entirety of `children`?

* When setting `capa_type` (or any future filter), will that continue to narrow down `children`, or will it narrow down `candidates` instead? (I presume the former, just want to be sure).

connorhaugh replied:
1. The former. Candidates tracks the "selected" blocks in the editor
2. For now, the former -- though I think filtering should be done on the frontend -- aka given a list of block Ids, let me ask the tagging service which ones of them have the given search terms.

kdmccormick replied:
@connorhaugh
1. Sounds good. FWIW, the other way would be OK too, I just want to make sure we know which we're doing.
2. 👍🏻

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -151,12 +155,29 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
         scope=Scope.user_state,
     )
+    shuffle = Boolean(
+        #is the block content randomized for every user.
```
```suggestion
        # Is the order of the blocks randomized for each learner?
```
let's be precise

connorhaugh replied:
` # Are the blocks seen and their order randomized for each learner?` is what i went with.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -251,6 +234,64 @@ def _publish_event(self, event_name, result, **kwargs):
+                if new_selected != old_selected:
+                    random.shuffle(new_selected)
+        else:
+            new_selected = list(valid_block_keys)[:size]
```
this would pick the first N blocks out of the library every single time, which ain't right.

instead, N blocks should be randomly picked from the library for each learner, and if `shuffle` is enabled, then the order of them should be randomized.

connorhaugh replied:
I'm confused by `this` here. 

if this refers to line 286:
note that line 286 provides selection for the case when shuffle = false.

if `this` refers to lines 267-284:

Note we need to preserve `selected` as much as possible, not across users, but rather for the same user. in cases where blocks are removed from the library, or max_count is decreased, we need to keep learners with all possible valid questions they have seen before. This way, updates to the block on publish will not vary significantly. The randomization from sample and shuffle is still included.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -251,6 +234,64 @@ def _publish_event(self, event_name, result, **kwargs):
+        return {(child.block_type, child.block_id) for child in library_children}
+
+    @classmethod
+    def make_selection(cls, selected, library_children, candidates, max_count, manual, shuffle):
```
This is already a complex method (perhaps the most complex and important method in the whole project), but the mixing of sets and lists, as well as the mixing of `.difference/.intersection` vs. `-/&`, is making it extra hard to follow.

Please:
* Convert to sets, unless order matters.
* Use type annotations (e.g., `set[tuple[str, str]]` and `list[tuple[str, str]]`)
* Use set operators (`&`, `-`, and `|`) rather than set methods (`intersection`, `difference`, and `union`).

connorhaugh replied:
What I decided here is that list might actually be the more useful data type for the for the following reasons:

We need to preserve ordering of the components across selections, and sets are unordered.
We need to do list operations many more times than set operations. Specifically, we need to truncate the selections by length and concatenate lists while preserving ordering. 

In summation, Although I am changing details, the ethos of your comment is being applied: make things consistent.

kdmccormick replied:
Thanks for switching from set methods to set operators; those do help.

> What I decided here is that list might actually be the more useful data type for the for the following reasons:

I wouldn't say that either lists or sets are more useful in this function; sorry if it seemed like I was nudging you to pick one or the other. They both have a place in this function. Converting everything to a list does not make the function more readable or correct. Conversely, it obscures which variables have a defined order.

There's a bug in new implementation that I want to use as an example. Look at valid_old_block_keys. Semantically, this should be a set, since its order is arbitrary (that is: neither well-defined nor truly random), but you're converting it to a list and then slicing that list in order to decide which keys to boot out:

```python
new_selected = valid_old_block_keys[0:size]
overlimit_block_keys = valid_old_block_keys[size:]
```

This will mean that the blocks will not get booted out randomly, as your docstring helpfully says they should be; rather, they'll be booted out at the whim of the order of valid_old_block_keys, which could be biased towards the blocks at the end of the `children` list, or perhaps the alphabetical order of the block keys, or something else (Python make no guarantees).

In the old function, this was done using sets, which meant they couldn't be sliced:
```python
overlimit_block_keys = set(rand.sample(selected_keys, num_to_remove))
selected_keys -= overlimit_block_keys
```

I realize that the logic of the new function is necessarily more complex (due to candidates being a thing now), but that's all the more reason to be careful with ordering.

Also, please please please add type annotations.

`old_selected: set = {tuple(k) for k in selected}`

vs

`old_selected: list = [tuple(k) for k in selected]`.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -485,6 +519,45 @@ def get_user_id(self):
+        """
+        Return source_block_ids for selection.
+        """
+        return Response(json.dumps({'source_block_ids': self.candidates}))
```
get_block_ids, selection, candidates...

please be consistent with terminology. These are not just any old block ids, and they are not the selection either. They are the candidates.

kdmccormick reviewed (changes requested):
Kinda hard to review without the rebase, but here are some comments for the meantime.

connorhaugh commented:
Taking a look at these first round of comments today, thanks so much.

connorhaugh committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

Commit:
fix: fixes from feedback and post-merge

Commit:
fix: lint

connorhaugh committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -105,23 +106,12 @@ class LibraryContentBlock(
-        if settings.FEATURES.get('MARK_LIBRARY_CONTENT_BLOCK_COMPLETE_ON_VIEW', False):
-            return XBlockCompletionMode.COMPLETABLE
-
-        return XBlockCompletionMode.AGGREGATOR
```
@connorhaugh are you intentionally removing completion_mode?

connorhaugh replied:
No idea how that was lost in the merge. adding back now...

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -264,6 +204,86 @@ def _publish_event(self, event_name, result, **kwargs):
+        4. In case b, we take the selected blocks that were valid before the edit,
+            and remove blocks randomly until he number of blocks is max_count.
+        5. In case c, we remove the blocks that are now invalid,
+           and supplament those remaining with new blocks chosen at random until the number of blocks is max_count.
```
Great details 🥇

connorhaugh replied:
Thanks -- I can't respond to the below comment comment for some reason, but I appreciate the thorough feedback and will take it and run with it.

connorhaugh replied:
In particular the type annotations. I need to get in the habit of using that tool.

kdmccormick reviewed (changes requested):
Partial pass. I'll come back finish the pass later today, hopefully.

Commit:
fix: selection fixes

Commit:
fix: lint

Commit:
fix: overlimit bug

Commit:
fix: unit tests + cleaner code

Commit:
fix: readd completion mode

connorhaugh committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -726,15 +790,18 @@ def source_library_values(self):
         """
         source_lib_changed = (self.source_library_id != old_metadata.get("source_library_id", ""))
         capa_filter_changed = (self.capa_type != old_metadata.get("capa_type", ANY_CAPA_TYPE_VALUE))
+        version_changed = (self.source_library_version != old_metadata.get("source_library_id", ""))
```
this is currently unused. did you mean to add it to the `if` statement on the next line?

connorhaugh replied:
removed.

kdmccormick committed:
fix selection pairing

* pairing w connor

* we did it

* fix: typo: sele->selection

* fix: extra colon

Commit:
fix: candidate typing

Commit:
fix: lint

connorhaugh committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

connorhaugh commented:
@kdmccormick  this is hopefully ready for a final review.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -159,12 +155,29 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
+        #is the block content only drawn from content which is in the candidates list?
+        default=False,
+        cope=Scope.settings,
+    )
```
We shouldn't add mysterious new settings to an actively-used component editor without any helptext or author-facing documentation. Furthermore, the add-candidates-by-usage-key-string interface here is really bad UX and I don't think we should ever expose it to general Studio users.

The interactive editor you and Ray are working on will be much nicer. I think that's how we ought to debut the static ref capability to general users. Do you agree?

If so, then for this PR, we could declare the new settings as `hidden=True`. They would sit at their default values, unused, until the new editor is turned on. We could still test that all the backend static ref machinery works locally by toggling on the new editor in our devstacks/tutors.

Finally, if all of that sounds good, what would we be able to remove from this PR? I think the JS supporting the legacy editor could go. Would the reordering support on the children page also go away, or is that still necessary even with the new editor?

connorhaugh replied:
> The interactive editor you and Ray are working on will be much nicer. I think that's how we ought to debut the static ref capability to general users. Do you agree?

The nascent plan I had was to merge this after we are ready to launch the new editor. I realize that we should probably make the old editor not notice these changes, in order to preserve the experience in the event that we have to turn off the new library block editor. Thanks for bringing up that point.

> If so, then for this PR, we could declare the new settings as hidden=True.

iff that doesn't hide them from the xblock api, then yes (I will check that).

> I think the JS supporting the legacy editor could go. Would the reordering support on the children page also go away, or is that still necessary even with the new editor?

That JS supports the "view library children" page, which we are not removing, because the new editor does not support the editing of children. Therefore it is necessary, for now.

kdmccormick replied:
@connorhaugh 

> The nascent plan I had was to merge this after we are ready to launch the new editor. I realize that we should probably make the old editor not notice these changes, in order to preserve the experience in the event that we have to turn off the new library block editor. Thanks for bringing up that point.
> iff that doesn't hide them from the xblock api, then yes (I will check that).

👍🏻 

> That JS supports the "view library children" page, which we are not removing, because the new editor does not support the editing of children. Therefore it is necessary, for now.

I understand we need to keep the "view library children" page, and that we need to make that page aware of `.candidates`, but I think the JS/HTML changes in this PR go beyond just that:

* library_content_block.js seems to support the legacy block picker, which will be unused.
* The edit to container.html which adds a "Save" button is there to support re-ordering, right? Do we actually want to support re-ordering here, or will the new component editor handle that?

connorhaugh replied:
> The edit to container.html which adds a "Save" button is there to support re-ordering, right?

No, it is there to support candidate selection on that page.

connorhaugh replied:
> library_content_block.js seems to support the legacy block picker, which will be unused.


i'll get rid of this stuff.

connorhaugh replied:
`non_editable_metadata_fields` is the  hidden = True you were looking for btw. It still appears in the xblock api, it just is not displayed in the editor

kdmccormick replied:
excellent 👍🏻

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -159,12 +155,29 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
         scope=Scope.user_state,
     )
+    shuffle = Boolean(
+        # Are the blocks seen and their order randomized for each learner?
```
```suggestion
        # Do we shuffle (randomize order) of selected blocks for each learner?
        # True -> Order is randomized for each learner.
        # False -> Original order from candidates/children is used.
```

connorhaugh replied:
Reasonable. I will add this.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -159,12 +155,29 @@ def completion_mode(cls):  # pylint: disable=no-self-argument
+        scope=Scope.settings,
+    )
+    manual = Boolean(
+        #is the block content only drawn from content which is in the candidates list?
```
```suggestion
        # Should selected blocks be limited to the manually-picked candidates?
        # True -> Draw selections from `candidates`.
        # False -> Draw selections from `children`.
```

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -382,10 +437,15 @@ def _get_selected_child_blocks(self):
         child_context = {} if not context else copy(context)
 
+        logger.error(f'hello world foralbe{self.get_children}')
+
```
```suggestion
```

connorhaugh replied:
So true.

Commit:
fix: hide shuffle, manual, candidates from editor.

Commit:
fix: result of manual testing

kdmccormick committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

kdmccormick commented:
@connorhaugh I merged in master. I'm also going to try making a PR sandbox. If you find that annoying, feel free to just take off the `create-sandbox` label.

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5763190794.log

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5763190794-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5763190794-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5763190794-tutor-requirements.txt

Please check the settings and requirements and retry deployment by updating the pull request description or pushing a new commit

gabor-boros commented:
☝️ This issue is caused by [Tutor](https://github.com/overhangio/tutor/blob/afb85aaab6b14344a2fdde548eb744ddc667b4fb/tutor/templates/apps/openedx/settings/partials/common_all.py#L153):

```
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 843, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/openedx/edx-platform/lms/envs/tutor/production.py", line 157, in <module>
    from django.utils.deprecation import RemovedInDjango50Warning, RemovedInDjango51Warning
ImportError: cannot import name 'RemovedInDjango50Warning' from 'django.utils.deprecation' (/openedx/venv/lib/python3.8/site-packages/django/utils/deprecation.py)
```

We are quickly checking if that was any discrepancy on our side when building the PR sandbox or if this in fact some problem with Tutor.

cc: @kdmccormick & @kaustavb12

kdmccormick commented:
Thanks for the heads up @gabor-boros . Let's discuss in Slack as to not disrupt this PR review.

connorhaugh commented:
All good @kdmccormick if we can get that spun up, that would be cool! No rush to merge this into master, as the new editor needs to go first.

connorhaugh committed:
feat: add new library editor toggle (#33628)

* feat: add new library editor toggle

* fix: small fix to html

* fix: lint

* fix: nit link from tnl ticket to issue

* fix: add back in tx listing page

* fix: add "content" to var names for clarity

kdmccormick commented on cms/templates/container.html:
```diff
@@ -176,7 +176,15 @@ <h3 class="sr">${_("Page Actions")}</h3>
                     </li>
+
                 % else:
+                    % if is_library_xblock:
+                    <li class="action-item action-save nav-item">
+                        <a href="#" class="button button-save action-button save-button is-hidden">
+                            <span class="action-button-text">${_("Save changes")}</span>
+                        </a>
+                    </li>
+                    % endif
```
I set up library_content block with manual==True and shuffle==False, went to the children page, and tried selecting children by checking boxes, but I don't see any "Save" button, so I can't save my selection.

![image](https://github.com/openedx/edx-platform/assets/3628148/e18df951-1e3c-4b9f-a019-eaae61c8012a)

connorhaugh replied:
It might be you have to remake static assets/refresh. There is some client side JS in the PR, but I will also take a look.

connorhaugh replied:
Confirmed. Updating static assets does the trick. `paver update_assets cms` is what i did in the cms shell.

kdmccormick replied:
Ohp, you're right!

kdmccormick reviewed (changes requested):
I think something's up with the children page. Were you able to get it working locally?

kdmccormick committed:
Merge branch 'master' into feat--Library-Content-Block-Reference

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928144.log

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928144-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928144-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928144-tutor-requirements.txt

Please check the settings and requirements and retry deployment by updating the pull request description or pushing a new commit

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928146-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928146-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5787928146-tutor-requirements.txt

rayzhou-bit committed:
fix: manual scope typo (#33957)

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5796328438-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5796328438-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5796328438-tutor-requirements.txt

kdmccormick commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -98,9 +99,21 @@ def transform_block_filters(self, usage_info, block_structure):
                 previous_count = len(selected)
-                block_keys = LibraryContentBlock.make_selection(selected, library_children, max_count, mode)
+                block_keys = LibraryContentBlock.make_selection(
+                    old_selected=list(map(tuple, selected)),
```
```suggestion
                    # the cache stores `selected` as list[list[str]], but we want to work with list[tuple[str, str]]
                    old_selected=[(block_type, block_id) for block_type, block_id in selected],
```
nit: comprehensions are considered more pythonic. Also, here's a suggested comment.

kdmccormick replied:
ack, pylint doesn't like this comprehension:

```
Unnecessary use of a comprehension, use list(self.selected) instead.
```

I'll put back what you had originally.

kdmccormick commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -98,9 +99,21 @@ def transform_block_filters(self, usage_info, block_structure):
                     if usage_key in library_children:
                         selected.append(selected_block)
 
+                #Get "canidate" blocks which have been manually selected
```
```suggestion
                # Get "candidate" blocks which have been manually picked by the author
```
NOT "SELECTED" 😄

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -436,15 +504,11 @@ def author_view(self, context):
                 context['can_edit_visibility'] = False
                 context['can_move'] = False
                 context['can_collapse'] = True
+                context['selectable'] = True
```
We should probably leave `selectable = False` when `manual==False`, do you agree?

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -436,15 +504,11 @@ def author_view(self, context):
-                    "library-block-author-preview-header.html", {
-                        'max_count': max_count,
-                        'display_name': self.display_name or self.url_name,
-                    }))
```
Removing this fragment would remove the message:

> Showing all matching content eligible to be added into My Static Library Content. Each student will be assigned $max_count components drawn randomly from this list.

That seems like a helpful message, why not keep it, but with extra parameter that lets you adjust it when `manual==True`?

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -436,15 +504,11 @@ def author_view(self, context):
                 context['can_move'] = False
                 context['can_collapse'] = True
+                context['selectable'] = True
                 self.render_children(context, fragment, can_reorder=False, can_add=False)
```
It would be awesome to set `can_reorder=True` when `shuffle==False`, but I realize there's some complexity around that, so I'm very OK with leaving `can_reorder=False` for now.

That said: Users can user re-orders the candidates in @rayzhou-bit 's editor, right? If so, is it correct that the order will be saved into candidates, but not children? If yes to both of those, then the order of this view may be wrong. I'm OK with leaving that as an issue for now, but I'd want to flag it for follow-up in the V2 effort.

kdmccormick reviewed (changes requested):
This is getting close. I'm stress-testing make_selection with a bunch of edge cases now; I'll let you know how that goes. In the meantime, here a few more comments, hopefully easy ones.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5805202913-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5805202913-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5805202913-tutor-requirements.txt

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
temp: feat: show the LibraryContentBlock's manual and shuffle fields

We should probably revert this, or at least put the fields behind a
flag, before merging, since we haven't prepared an announcement for
how to use these new fields, which would appear for all
LibraryContentBlock users.

kdmccormick commented:
@connorhaugh Despite what I said earlier about hiding manual and shuffle, I've been finding it really hard to test this without those fields available in Studio 😛

So, I added the fields back in this commit, plus some help text: https://github.com/openedx/edx-platform/pull/33511/commits/6edfd51f05e4add37be033d912c98428fc2fad9f . With that commit in place, we can use the sandbox on this PR to test out static reference.

You'll *probably* want to revert that commit before merging this, but honestly, up to you and Brad. I'm no longer concerned about how the change will look in Redwood, since I know that we'll have resources to clean up the UI before the release is cut.

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042755.log

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042755-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042755-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042755-tutor-requirements.txt

Please check the settings and requirements and retry deployment by updating the pull request description or pushing a new commit

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042752-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042752-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5806042752-tutor-requirements.txt

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5811369328-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5811369328-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5811369328-tutor-requirements.txt

kdmccormick committed:
test: type-check library_content_block.py

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5812805645-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5812805645-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5812805645-tutor-requirements.txt

kdmccormick committed:
test: beef up the unit tests for randomized LibraryContentBlock a bit

kdmccormick committed:
test: run a test that wasn't running. it's failing now

kdmccormick commented on xmodule/tests/test_library_content.py:
```diff
@@ -736,3 +739,130 @@ def test_removed_invalid(self):
+        self.lc_block.manual = True
+        self._bind_course_block(self.lc_block)
+
+    def test_selects_children_from_candidates(self):
```
This test had a few issues, including that its name didn't start with `test_`, so it wasn't running. I've fixed a couple things up, but its assertion is still currently failing.

kdmccormick replied:
Fixed in https://github.com/openedx/edx-platform/pull/33511/commits/675fed6397ac90cb8ced7269b932651fa4e62cc1

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813248768-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813248768-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813248768-tutor-requirements.txt

kdmccormick committed:
fix: pass candidates into make_selection correctly

fixes three bugs:

* all candidates were being filtered out during tests, because
  (until saved) self.candidates contains tuples, not nested lists.
* children order was being used instead of candidates order.
* empty list of candidates should yield zero blocks, not all blocks.

kdmccormick committed:
test: test shuffled & unshuffled variant of all selection tests

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813741212-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813741212-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5813741212-tutor-requirements.txt

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
fix: bug and pylint violations in make_selection calls

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828035825-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828035825-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828035825-tutor-requirements.txt

kdmccormick committed:
test: LibraryContentBlock test_invalid_block_without_replacement

kdmccormick committed:
test: complex library content selection scenario (currently failing)

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

open-craft-grove commented:
Sandbox deployment successful 🚀 

 🎓 [LMS](https://pr-33511-139931.sandboxes.opencraft.hosting)
 📝 [Studio](https://studio.pr-33511-139931.sandboxes.opencraft.hosting)
 ℹ️ [Tutor config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828405462-tutor-config.yml), [Tutor requirements](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828405462-tutor-requirements.txt), [Grove config](https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5828405462-grove-config.yml)

kdmccormick committed:
test: make test output easier to read & remove breakpoint

kdmccormick committed:
test: don't just convert from lists to sets in tests (causes flakiness)

kdmccormick committed:
test: stabilize LCB shuffling+sampling in tests (complex is still flaky)

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5848332929-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5848332929-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5848332929-tutor-requirements.txt

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
test: making selection tests lms-only (wip)

rayzhou-bit committed:
feat: include children info in xblock api (#33977)

kdmccormick committed:
test: selection tests are working, with a hack

kdmccormick committed:
Merge branch 'feat--Library-Content-Block-Reference' of github.com:openedx/edx-platform into feat--Library-Content-Block-Reference

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
test: continue skipping lc-sync tests in LMS

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855131079-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855131079-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855131079-tutor-requirements.txt

kdmccormick committed:
test: fix cms tests

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855306666-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855306666-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5855306666-tutor-requirements.txt

kdmccormick committed:
test: better var names in selection tests

kdmccormick committed:
refactor: 'valid_children' -> 'available'

kdmccormick committed:
test: selection tests work in both manual and non-manual modes now

kdmccormick committed:
test: attempt at making selection tests run quicker with sharedmodulestore

kdmccormick committed:
revert: test: attempt at making selection tests run quicker with sharedmodulestore

This reverts commit 0f83fcf7466d7d84fd5f9b9b6a963dcf5fa7e4a7.

kdmccormick committed:
test; add docstring to a test helper

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5856406501-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5856406501-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5856406501-tutor-requirements.txt

kdmccormick committed:
test: make selection test setup more verbose but also more straightforward

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
test: better names and comments for selection tests

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862306046-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862306046-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862306046-tutor-requirements.txt

kdmccormick committed:
test: in selection tests, test more behavior with fewer cases

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -264,6 +242,108 @@ def _publish_event(self, event_name, result, **kwargs):
         self.runtime.publish(self, f"edx.librarycontentblock.content.{event_name}", event_data)
         self._last_event_result_count = len(result)  # pylint: disable=attribute-defined-outside-init
 
+    def available_children(self) -> list[tuple[str, str]]:
```
I exposed this as a public method so that unit tests can use it without duplicating the same logic.

The underlying implementation is still a static method so that make_selection can use it.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -264,6 +242,108 @@ def _publish_event(self, event_name, result, **kwargs):
+        )
+
+    @classmethod
+    def _available_children(
```
This was previously "_get_valid_children". I renamed it "_available_children" because it struck me as more self-explanatory -- these are the children which are available for selection.

kdmccormick commented on xmodule/library_content_block.py:
```diff
@@ -264,6 +242,108 @@ def _publish_event(self, event_name, result, **kwargs):
+        if manual:
+            return [candidate for candidate in candidates if candidate in children]
+        else:
+            return children
```
This is one of the recent "bugs" I fixed, if you'd call it that.

Before, when `manual==True`, we would return the whole candidates list, regardless of whether all of those candidates appeared as block children.

Now, when `manual=True`, we return the candidates *which are also children*. That way, if (for some reason) child is removed but the candidate entry is not, we don't get `ItemNotFound` errors from modulestore. I'm not sure when that issue would happen, but it seems like a reasonable safeguard to have in place.

kdmccormick reviewed (commented):
@connorhaugh Just an FYI on some recent changes I've pushed.

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862781415-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862781415-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5862781415-tutor-requirements.txt

kdmccormick committed:
test: add cases for: no selection available, toggling manual on/off

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5863446981-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5863446981-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5863446981-tutor-requirements.txt

Commit:
fix: new editor + candidates_in_course

open-craft-grove commented:
Sandbox deployment failed.

Check failure logs here https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898626.log

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898626-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898626-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898626-tutor-requirements.txt

Please check the settings and requirements and retry deployment by updating the pull request description or pushing a new commit

open-craft-grove commented:
Sandbox deployment successful.

Sandbox LMS is available at https://pr-33511-139931.sandboxes.opencraft.hosting
Sandbox Studio is available at https://studio.pr-33511-139931.sandboxes.opencraft.hosting

Instance Tutor config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898607-tutor-config.yml
Instance Grove config : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898607-grove-config.yml
Instance Python requirements : https://axim-sandbox-build-logs-4626178240.nyc3.digitaloceanspaces.com/pr-33511-139931-51564656-5872898607-tutor-requirements.txt

kdmccormick committed:
fix: candidates are always library block usage keys (WIP)

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
fix: derive_key import and usage in LCB

kdmccormick committed:
fix: candidates are UsageKeys from the source library; tests still failing

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick committed:
Merge remote-tracking branch 'upstream/master' into feat--Library-Content-Block-Reference

kdmccormick commented:
Closing, as we will be implementing static reference on a per-component basis rather than though the LibraryContentBlock.
